lib/ReadsUtils/ReadsUtilsImpl.py

```diff
@@ -82,9 +87,20 @@ def _check_pos(self, num, name):
     def _proc_upload_reads_params(self, params):
         fwdid = params.get('fwd_id')
         fwdfile = params.get('fwd_file')
-        if not self.xor(fwdid, fwdfile):
-            raise ValueError('Exactly one of a file or shock id containing ' +
+        fwdurl = params.get('fwd_file_url')
+        fwdstaging = params.get('fwd_staging_file_name')
+        if sum(bool(e) for e in [fwdid, fwdfile, fwdurl, fwdstaging]) != 1:
+            raise ValueError('Exactly one of a file, shock id, staging file name or file url containing ' +
                              'a forwards reads file must be specified')
+        if fwdurl:
+            download_type = params.get('download_type')
+            if not (fwdurl and download_type):
+                raise ValueError('Both download_type and fwd_file_url must be provided')
+            fwdfile = self._download_web_file(fwdurl, download_type)
+        
+        if fwdstaging:
+            fwdfile = self._download_staging_file(params.get('user_id'), fwdstaging)
+
```

Comment:
My feeling is that all the download code, whether from shock or elsewhere, should be consolidated in one part of the fn (using DFU as appropriate), and any errors thrown should identify the source of the data